### PR TITLE
CSP-1617: Extended SendNotificationsFunction to handle both Provider and Employer scenarios

### PR DESCRIPTION
**Pull request by shaun-downey-education**
- Updated the **SendNotificationsFunction** to handle both Employer and Provider journeys.

### CODE REVIEW

File: src/SFA.DAS.PR.Jobs/Functions/Notifications/SendNotificationsFunction.cs [hunk 1]
```diff
@@ -64,13 +70,29 @@ public async Task Run([TimerTrigger("%SendNotificationsFunctionSchedule%", RunOn
...
         try
         {
-            ProviderEmailRequest providerEmailRequest = await CreateProviderEmailRequest(notification, cancellationToken);
-            await _pasAccountApiClient.SendEmailToAllProviderRecipients(notification.Ukprn!.Value, providerEmailRequest, cancellationToken);
-            notification.SentTime = DateTime.UtcNow;
+            switch(notification.NotificationType)
+            {
+                case nameof(NotificationType.Provider):
+                    {
+                        ProviderEmailRequest providerEmailRequest = await CreateProviderEmailRequest(notification, cancellationToken);
```
Comment (PeterDavidWilliams) [inline src/SFA.DAS.PR.Jobs/Functions/Notifications/SendNotificationsFunction.cs:82]: This request object and the corresponding employer request object (ProviderEmailRequest and SendEmailCommand) are differently formatted and it feels a bit odd

**Review by PeterDavidWilliams — COMMENTED** [review]
I've looked at what's already here, left one comment. I'll take another look once you've covered the edge cases you've mentioned on slack

File: src/SFA.DAS.PR.Jobs/Functions/Notifications/SendNotificationsFunction.cs (hunk 1 above)
Comment (shaun-downey-education) [reply src/SFA.DAS.PR.Jobs/Functions/Notifications/SendNotificationsFunction.cs:82]: @PeterDavidWilliams Yea one goes off to the notifications message queue and one goes off to an external API, that's just the format of the required objects so no issue there.

Comment (PeterDavidWilliams) [reply src/SFA.DAS.PR.Jobs/Functions/Notifications/SendNotificationsFunction.cs:82]: Ok, happy then, resolved

File: src/SFA.DAS.PR.Data/Common/EmailTokens.cs
Comment (DevenKShah) [inline src/SFA.DAS.PR.Data/Common/EmailTokens.cs:1]: This should be renamed to NotificationTokens and the const can avoid Token prefix as they will be suffixed by the class name any way.

File: src/SFA.DAS.PR.Jobs/Functions/Notifications/SendNotificationsFunction.cs [hunk 2]
```diff
@@ -64,13 +70,29 @@ public async Task Run([TimerTrigger("%SendNotificationsFunctionSchedule%", RunOn
         _logger.LogInformation("{FunctionName} - Processed {ProcessedCount} notifications.", nameof(SendNotificationsFunction), processedCount);
     }
 
-    private async Task<int> TryProcessProviderNotification(Notification notification, CancellationToken cancellationToken)
+    private async Task<int> TryProcessProviderNotification(Notification notification, FunctionContext executionContext, CancellationToken cancellationToken)
```
Comment (DevenKShah) [inline src/SFA.DAS.PR.Jobs/Functions/Notifications/SendNotificationsFunction.cs]: This method should be renamed now as it processes both provider and employer notifications

File: src/SFA.DAS.PR.Jobs/Functions/Notifications/SendNotificationsFunction.cs [hunk 3]
```diff
@@ -64,13 +70,29 @@ public async Task Run([TimerTrigger("%SendNotificationsFunctionSchedule%", RunOn
...
+                case nameof(NotificationType.Employer):
+                    {
+                        SendEmailCommand employerEmailRequest = await CreateEmployerEmailRequest(notification, cancellationToken);
+
+                        await _functionEndpoint.Send(employerEmailRequest, executionContext, cancellationToken);
+                    }
+                    break;
+            }
+
+            await UpdateNotification(notification, cancellationToken);
```
Comment (DevenKShah) [inline src/SFA.DAS.PR.Jobs/Functions/Notifications/SendNotificationsFunction.cs:102]: We need to commit each notification as it is sent out. Currently if any one fails all of the notifications will be retried, which will mean that we will send out duplicate notifications.

**Review by DevenKShah — CHANGES_REQUESTED** [review]

**Review by DevenKShah — DISMISSED** [review]

**Review by PeterDavidWilliams — APPROVED** [review]

**Review by furqanagwan — APPROVED** [review]

**Pull request merged**